apollo-coroutines-support/src/main/kotlin/com/apollographql/apollo/coroutines/CoroutinesExtensions.kt

```diff
@@ -73,35 +73,33 @@ fun <T> ApolloQueryWatcher<T>.toFlow(): Flow<Response<T>> = callbackFlow {
 
 
 /**
- * Converts an [ApolloCall] to an [Deferred]. This is a convenience method that will only return the first value emitted.
- * If the more than one response is required, for an example to retrieve cached and network response, use [toFlow] instead.
+ * Suspends the [ApolloCall] until it completes and returns the value on success or throws an exception on failure.
+ * The [ApolloCall] is cancelled when the coroutine running the operation is cancelled as well.
  *
  * @param <T>  the value type.
- * @return the deferred
+ * @return the response on success.
+ * @throws ApolloException on failure.
  */
-fun <T> ApolloCall<T>.toDeferred(): Deferred<Response<T>> {
-  val deferred = CompletableDeferred<Response<T>>()
-
-  deferred.invokeOnCompletion {
-    if (deferred.isCancelled) {
-      cancel()
-    }
-  }
+suspend fun <T> ApolloCall<T>.await(): Response<T> = suspendCancellableCoroutine { cont ->
   enqueue(object : ApolloCall.Callback<T>() {
+
+    private var resumeIsCalled = false
+
     override fun onResponse(response: Response<T>) {
-      if (deferred.isActive) {
-        deferred.complete(response)
+      if (!resumeIsCalled) {
+        cont.resume(response)
+        resumeIsCalled = true
       }
     }
 
     override fun onFailure(e: ApolloException) {
-      if (deferred.isActive) {
-        deferred.completeExceptionally(e)
-      }
+      cont.resumeWithException(e)
```

Comment:
You should also guard here too I guess since you can possibly get multiple failures.

Reply:
(or one success and one failure)

Reply:
From the documentation it says `onFailure` can be called at most one time, so I don't think another guard is needed here. 
https://github.com/apollographql/apollo-android/blob/9513025a9640517f0fa448c9859938106c50606a/apollo-runtime/src/main/java/com/apollographql/apollo/ApolloCall.java#L103

Reply:
Ah ok got it. What about the one failure and n successes? Can that happen?

Reply:
The documentation says, that onFailure is a terminal that means if it happens, onSuccess will never be called, however I'm not sure if we can get onSuccess prior the onFailure or not. Better to hear from the maintainers in that case.

Reply:
Got it thanks for clarification! I'll leave up to to you guys who know more

Reply:
`onSuccess` will be called several time if using the cache (one for the cache response, another for the network response for an exemple). If fetching from the network fails, there will be one `onSuccess` and one `onFailure` call. 

Users needing to use cache responses should really use`toFlow` instead but `await()` should not crash if that ever happens.

Also, I think there's no guarantee what thread the callbacks are called in so `resumeIsCalled` would have to be thread-safe. Maybe something like:

```kotlin
resumeIsCalled = AtomicBoolean()
if (!resumeIsCalled.getAndSet(true)) {
   cont.resume(response)
}
```